virtualizarr/manifests/store.py

```diff
@@ -137,34 +133,92 @@ def parse_manifest_index(key: str, chunk_key_encoding: str = ".") -> tuple[int,
 
 
-def find_matching_store(stores: StoreDict, request_key: str) -> StoreRequest:
-    """
-    Find the matching store based on the store keys and the beginning of the URI strings,
-    to fetch data from the appropriately configured ObjectStore.
+def _find_bucket_region(bucket_name: str) -> str:
+    import requests
 
-    Parameters:
-    -----------
-    stores : StoreDict
-        A dictionary with URI prefixes for different stores as keys
-    request_key : str
-        A string to match against the dictionary keys
+    resp = requests.head(f"https://{bucket_name}.s3.amazonaws.com")
+    return resp.headers["x-amz-bucket-region"]
 
-    Returns:
-    --------
-    StoreRequest
+
+def _default_object_store(filepath: str) -> tuple[str, ObjectStore]:
+    import obstore as obs
+
+    parsed = urlparse(filepath)
+
+    if parsed.scheme in ["", "file"]:
+        return "file://", obs.store.LocalStore()
+    if parsed.scheme == "s3":
+        bucket = parsed.netloc
+        return f"s3://{bucket}", obs.store.S3Store(
+            bucket=bucket,
+            client_options={"allow_http": True},
+            skip_signature=True,
+            virtual_hosted_style_request=False,
+            region=_find_bucket_region(bucket),
+        )
+
+    raise NotImplementedError(f"{parsed.scheme} is not yet supported")
+
+
+def _sort_stores_by_prefix_length(input_dict):
+    sorted_items = sorted(input_dict.items(), key=lambda x: len(x[0]), reverse=True)
+    return dict(sorted_items)
+
+
+class ObjectStoreRegistry:
+    """
+    ObjectStoreRegistry maps URLs to ObjectStore instances, and allows ManifestStores to read from different ObjectStore instances.
     """
-    # Sort keys by length in descending order to ensure longer, more specific matches take precedence
-    sorted_keys = sorted(stores.keys(), key=len, reverse=True)
-
-    # Check each key to see if it's a prefix of the uri_string
-    for key in sorted_keys:
-        if request_key.startswith(key):
-            parsed_key = urlparse(request_key)
-            return StoreRequest(store=stores[key], key=parsed_key.path)
-    # if no match is found, raise an error
-    raise ValueError(
-        f"Expected the one of stores.keys() to match the data prefix, got {stores.keys()} and {request_key}"
-    )
+
+    _stores: dict[str, ObjectStore]
+
+    @classmethod
+    def __init__(self, stores: dict | None = None):
+        stores = stores or {}
+        for store in stores.values():
+            if not store.__class__.__module__.startswith("obstore"):
+                raise TypeError(f"expected ObjectStore class, got {store!r}")
+        self._stores = _sort_stores_by_prefix_length(stores)
+
+    def register_store(self, url: str, store: ObjectStore):
+        """
+        Register a store using the given url
+
+        If a store with the same key existed before, it is replaced
+        """
+        self._stores[url] = store
+        self._stores = _sort_stores_by_prefix_length(self._stores)
```

Comment:
This means that every time you insert a store, you sort the full list of stores..?

I think having a sequence as an internal data structure makes sense when the user passes in a list, because they can control the order of stores within that list.

But when the user adds and removes stores via e.g. `register_store`, I think a HashMap makes more sense under the hood (and is closer to what `DataFusionRegistry` is doing. It uses `dashmap`, a concurrent hashmap)